### DISCUSSION TITLE
Markdown Editor Styling Overwritten by @shadcn/ui Typography in React App

### DISCUSSION THREAD

**Discussion by softshipper** [shadcn-ui/ui#4051] [Q&A]
Hi there,

I’m working on a React application that uses @shadcn/ui for the UI framework and react-markdown for rendering Markdown content. My problem is that the Markdown content is not displaying correctly because the typography styles from @shadcn/ui are overriding the Markdown styles. Specifically, the Markdown headers and other text elements are not showing the expected formatting.

Here is the relevant part of my React component:

    import React, { useEffect, useState } from "react"
    import { MarkdownFile, isHTTPValidationError } from "@/types"
    import { useParams, useNavigate } from "react-router-dom"
    import { fetchMarkdown, updateMarkdown } from "@/endpoint"
    import { Button } from "./ui/button"
    import Markdown from "react-markdown"
    import rehypeHighlight from "rehype-highlight"
    import remarkGfm from "remark-gfm"
    import "./styles.css"
    
    const MarkdownEdit: React.FC = () => {
      const [markdown, setMarkdown] = useState<MarkdownFile | null>(null)
      const { id } = useParams<{ id: string }>()
      const navigate = useNavigate()
      const [error, setError] = useState<string>("")
    
      useEffect(() => {
        if (id) {
          fetchMarkdown(id, setMarkdown)
        }
      }, [id])
    
      const handleSave = async (e: React.FormEvent) => {
        e.preventDefault()
        setError("")
        if (!markdown) return
        try {
          await updateMarkdown(markdown.id, markdown, setMarkdown)
          navigate("/")
        } catch (err: unknown) {
          if (isHTTPValidationError(err)) {
            setError((err as Error & { detail: string }).detail)
          } else {
            setError("An unknown error occurred.")
          }
        }
      }
    
      return (
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            minHeight: "100vh",
            backgroundColor: "#F9FAFB",
          }}
        >
          {markdown && (
            <div
              style={{
                display: "flex",
                flexDirection: "column",
                height: "calc(100vh - 100px)", // Adjust this value as needed
                overflow: "hidden", // Prevents the page from scrolling
              }}
            >
              <h2
                style={{
                  marginBottom: "0.5rem",
                  fontSize: "1.875rem",
                  fontWeight: "600",
                }}
              >
                Company: {markdown.company_name}
              </h2>
              <h4 style={{ fontSize: "1.25rem", fontWeight: "600" }}>
                Business: {markdown.business_type}
              </h4>
              {error && <div className="text-red-500 mb-4">{error}</div>}
              <div
                style={{
                  display: "flex",
                  flexDirection: "row",
                  height: "100%", // Take up remaining space
                  overflow: "hidden", // Prevents this div from scrolling
                }}
              >
                <textarea
                  style={{
                    flex: 1,
                    width: "50%",
                    overflowY: "auto",
                    resize: "none",
                  }}
                  placeholder="Type your message here."
                  value={markdown.products}
                  onChange={(e) =>
                    setMarkdown({ ...markdown, products: e.target.value })
                  }
                />
                <div
                  className="markdown-content"
                  style={{
                    flex: 1,
                    width: "50%",
                    overflowY: "auto",
                    padding: "0.5rem",
                  }}
                >
                  <Markdown rehypePlugins={[rehypeHighlight, remarkGfm]}>
                    {markdown.products}
                  </Markdown>
                </div>
              </div>
            </div>
          )}
          <Button style={{ margin: "1rem" }} onClick={handleSave}>
            Save
          </Button>
        </div>
      )
    }
    
    export default MarkdownEdit

To solve this issue, I created a custom CSS class to override the global typography styles from @shadcn/ui. Here is the CSS:

    /* styles.css */
    .markdown-content h1,
    .markdown-content h2,
    .markdown-content h3,
    .markdown-content h4,
    .markdown-content h5,
    .markdown-content h6 {
      font-size: inherit;
      font-weight: inherit;
      line-height: inherit;
    }
    
    .markdown-content p {
      margin-top: inherit;
      margin-bottom: inherit;
      font-size: inherit;
      line-height: inherit;
    }
    
    .markdown-content blockquote {
      font-size: inherit;
      line-height: inherit;
      margin: inherit;
      padding: inherit;
      border-left: inherit;
      color: inherit;
    }
    
    .markdown-content ul,
    .markdown-content ol {
      list-style: inherit;
      margin: inherit;
      padding: inherit;
    }
    
    .markdown-content table {
      width: inherit;
      border-collapse: inherit;
      margin: inherit;
    }

By wrapping the Markdown component in a div with the markdown-content class, I ensured the Markdown styles were correctly applied:

    <div className="markdown-content">
      <Markdown rehypePlugins={[rehypeHighlight, remarkGfm]}>
        {markdown.products}
      </Markdown>
    </div>

Unfortunately, the header still doesn't show the correct format in the Markdown component.
[![enter image description here][1]][1]

  [1]: https://i.sstatic.net/pBmXSaBf.png
  
  Best regards

**Comment by HichemTab-tech** [comment]
Same here :")

**Comment by abhayclasher** [comment]
Hey! The issue is that shadcn's base typography styles (like `h1`, `h2`) are bleeding into your markdown container.\n\nInstead of writing all that manual CSS with `inherit`, the easiest fix is to use the `@tailwindcss/typography` plugin. \n\n1. `npm install -D @tailwindcss/typography`\n2. Add it to your `tailwind.config.js` plugins: `plugins: [require('@tailwindcss/typography')]`\n3. Slap the `prose` class on your markdown wrapper:\n\n```tsx\n<div className=\"prose dark:prose-invert max-w-none\">\n  <Markdown rehypePlugins={[rehypeHighlight, remarkGfm]}>\n    {markdown.products}\n  </Markdown>\n</div>\n```\n\nThis automatically scopes all the markdown styling correctly and stops shadcn's base styles from messing it up! 👍

**Comment by DaCameraGirl** [comment]
This is a very common conflict when mixing shadcn/ui (which uses Tailwind's Preflight/base reset) with markdown renderers. Here's the root cause and all the viable fixes.

## Root Cause

Tailwind's base styles (Preflight) reset all heading styles, list styles, etc. to browser defaults. shadcn/ui adds its own typography globals on top. When `react-markdown` renders `<h1>`, `<h2>`, `<ul>`, etc., those elements get the reset/global styles rather than the semantic markdown styles you expect.

## Fix 1: `@tailwindcss/typography` (Recommended)

The cleanest solution — the `prose` class applies a curated set of typographic styles specifically designed for HTML content rendered from Markdown or rich text:

```bash
npm install -D @tailwindcss/typography
```

```js
// tailwind.config.js
module.exports = {
  plugins: [require('@tailwindcss/typography')],
}
```

```tsx
<div className="prose dark:prose-invert max-w-none">
  <Markdown
    remarkPlugins={[remarkGfm]}
    rehypePlugins={[rehypeHighlight]}
  >
    {markdown}
  </Markdown>
</div>
```

Key classes:
- `prose` — applies the base typographic styles
- `dark:prose-invert` — inverts for dark mode
- `max-w-none` — removes the default max-width constraint (useful inside a container)
- `prose-sm`, `prose-lg`, `prose-xl` — size variants

## Fix 2: Scoped CSS Module

If you want full control without the `prose` plugin:

```css
/* MarkdownContent.module.css */
.markdown h1 { font-size: 2em; font-weight: bold; margin-bottom: 0.5em; }
.markdown h2 { font-size: 1.5em; font-weight: bold; margin-bottom: 0.4em; }
.markdown h3 { font-size: 1.25em; font-weight: semibold; margin-bottom: 0.3em; }
.markdown ul { list-style-type: disc; padding-left: 1.5em; }
.markdown ol { list-style-type: decimal; padding-left: 1.5em; }
.markdown p { margin-bottom: 1em; }
.markdown code { background: #f4f4f4; padding: 0.2em 0.4em; border-radius: 3px; }
.markdown pre { background: #1e1e1e; padding: 1em; border-radius: 6px; overflow-x: auto; }
.markdown blockquote { border-left: 4px solid #e5e7eb; padding-left: 1em; color: #6b7280; }
```

```tsx
import styles from './MarkdownContent.module.css'

<div className={styles.markdown}>
  <Markdown>{markdown}</Markdown>
</div>
```

## Fix 3: Custom Components via `react-markdown`

For maximum control, override specific elements directly in `react-markdown`:

```tsx
<Markdown
  remarkPlugins={[remarkGfm]}
  rehypePlugins={[rehypeHighlight]}
  components={{
    h1: ({ children }) => (
      <h1 className="text-3xl font-bold mb-4 mt-6 border-b pb-2">{children}</h1>
    ),
    h2: ({ children }) => (
      <h2 className="text-2xl font-semibold mb-3 mt-5">{children}</h2>
    ),
    h3: ({ children }) => (
      <h3 className="text-xl font-medium mb-2 mt-4">{children}</h3>
    ),
    ul: ({ children }) => (
      <ul className="list-disc list-inside mb-4 space-y-1">{children}</ul>
    ),
    ol: ({ children }) => (
      <ol className="list-decimal list-inside mb-4 space-y-1">{children}</ol>
    ),
    p: ({ children }) => <p className="mb-4 leading-7">{children}</p>,
    code: ({ inline, className, children }) =>
      inline ? (
        <code className="bg-muted px-1.5 py-0.5 rounded text-sm font-mono">
          {children}
        </code>
      ) : (
        <pre className="bg-muted p-4 rounded-lg overflow-x-auto">
          <code className={className}>{children}</code>
        </pre>
      ),
  }}
>
  {markdown}
</Markdown>
```

This approach lets you use shadcn's own design tokens (`bg-muted`, etc.) for a perfectly consistent look.

## Recommendation

| Situation | Best fix |
|---|---|
| Quick fix, minimal setup | Fix 1: `prose` class |
| Need design token consistency with shadcn | Fix 3: custom components |
| Complex/custom styles without plugin dependency | Fix 2: CSS module |

For most projects, **Fix 1 + Fix 3 combined** works best: use `prose` for baseline typographic rhythm, then override specific elements with shadcn tokens where needed.
